README.md
# Sarah’s Scala Utilities #

This project contains miscellaneous Scala utility code, mostly for my own
projects.  It contains components that are relatively independent and that are
likely to be reused without modification across a wide variety of projects.

Some code may be included that is only useful in specialized situations if has
minimal dependencies, is likely to be well contained, and has the generality
of utility code.

No guarantees are offered, but feel free to use this if it is helpful to you.
If you find any bugs, please send me a message or a pull request.  Like many
utility libraries, this code is *not* meant to be abuse proof.  If you try to
do something that seems like it might be a bad idea, you're on your own. :-)

Because of dependencies, this requires Scala 2.11 or higher. Java 8 is
required for version 4.0 or higher.

## Installation ##

    libraryDependencies += "org.gerweck.scala" %% "gerweck-utils" % "5.0.1"


### Akka Utilities ###

For utilities related to Akka and Akka Streaming, use this include. The Akka
utilities require Java 8 and Akka 2.4.

    libraryDependencies += "org.gerweck.scala" %% "gerweck-utils-akka" % "5.0.1"

### DB Utilities ###

Some DB lightweight utilities, based on pure JDBC, are included in the core
module. However, advanced utilties or those requiring libraries are in a
separate module. These modules require Java 8 and can be included as such:

    libraryDependencies += "org.gerweck.scala" %% "gerweck-utils-db" % "5.0.1"

## Contributors

- [Sarah Gerweck](https://github.com/sarahgerweck/) (primary author)
- [Rouzbeh Safaie](https://github.com/rsafaie/)


CHANGELOG.md
# Gerweck Utils Changelog

## 2.0

- Move core utilities into a `core` subfolder
  - Does not affect users of the library
- Add Akka submodule for actor & stream utils
- Drop support for Scala 2.10
- Default package now requires Java 8
  - For Java 6, use the `-java6` suffix
  - The old `-java8` suffix will no longer exist

### 2.1

- Add `prefs` and `mapping` packages
  - `Prefs` makes `java.util.prefs` more Scala friendly
  - `Mapping` is experimental, contains type morphisms

### 2.2

- Add `gerweck-util-db` submodule with DB utilities

#### 2.2.2

- Update to Akka 2.4.8
- Disable `-optimize` for Akka
  - Optimization with Akka tends to pin you to the exact version of Akka,
    which is not something we want for utilities.

#### 2.2.4

- Update to Akka 2.4.11

#### 2.2.5

- Make `VersionNumber` serializable

#### 2.2.6

- Update the Json4s optional dependency to 3.5.0.
- Update the Twitter Core optional dependency to 6.38.0.

### 2.3.0

- Add support for Scala 2.12
  - Slick is only available as a milestone release for 2.12, so you may not
    want to use the 2.12 version for production software if you use the
    Slick support.
- Update to Akka 2.4.14
- Update to Akka HTTP 10.0
  - Akka HTTP is no longer experimental, and it's now on a different release
    schedule from Akka
- Start dropping build support for Scala 2.12 milestones.
- Slight performance improvements in `TokenParserUtil`.

### 2.4.0

- Update to Akka 2.4.17
- Add `ZipStream` to Akka module, which provides a streaming, non-blocking
  mechanism for producing zip archives.
- Add `hashing` package that provides idiomatic access to hash algorithms.
  - This also adds support for the Bouncy Castle hash implementations. This
    is now an optional dependency.

### 2.5.0

- Updates to the `ZipStream` module, *including some breaking changes*.
- Deprecate the `foldl1` and `rx` extension methods. These are both
  available in recent version of the Scala standard library.
- Update to Akka 2.5.3
- Update to Akka HTTP 10.0.9
- Update to Slick 3.2

#### 2.5.1

- Update Scala to 2.12.3
- Add `Throwing` JDBC objects

### 2.6.0

- Update to SBT 1.0
- Minor dependency updates
- Update Scala to 2.12.4
- Add support for `java.nio.file.Path` preferences

### 2.7.0

- Update the `Prefs` API to allow for optional preferences
- Make Commons-IO and Commons-VFS optional dependencies
  - These are only used by the deprecated `FileUtils` system
- Make JCL-over-SLF4J an optional dependency
  - This should have always been optional.
  - I *strongly* recommend against using commons-logging, but there's no
    reason this library should make that decision for you.

#### 2.7.1

- Update Twitter Util to 17.12
- Update Akka to 2.5.8
- Update Akka HTTP to 10.0.11

#### 2.7.2

- Update SBT to 1.1
- Update BouncyCastle to 1.59

## 3.0

- Remove all deprecated interfaces
- Update Scala Parser Combinators to 1.1.0
- Update Akka to 2.5.9

#### 3.0.1

- Fix a bug that affects the `equals` of `VersionNumber` and add tests
- Dependency updates. (These are all binary compatible.)
  - Update Akka to 2.5.11
  - Update Scala XML to 1.1.0
  - Update Liquibase to 3.5.5
  - Update Log4s to 1.6.0
  - Update SBT to 1.1.1
  - Update Slick to 3.2.2
  - Update Akka HTTP to 10.1.0
  - Update Twitter Util to 18.3.0
- Added `unwrapOption` extension to `java.sql.Wrapper` classes
  - This is a more convenient way to do unwrapping in Scala
  - Import `org.gerweck.scala.util.jdbc._` to make it available
- Remove some dead code from Scala 2.10 & simplify the build

#### 3.0.2

- Build Scala 2.11 version with Scala 2.11.12
- Build Scala 2.12 version with Scala 2.12.5
  - Scala 2.12.5 includes some substantial performance improvements for string
    interpolation, so this may improve performance in some situations.

### 3.1

- Add SHAKE-128 and SHAKE-256 hash methods
- New `RichByteArray` extension methods
  - `toHexString`
  - `toBasicBase64`
  - `toUriBase64`

### 3.2

- Update Slick to 3.2.3
  - This is binary compatible with Slick 3.2.1 and 3.2.2.
- Update Akka HTTP to 10.1.1
- Update Liquibase to 3.6.0
- Non-functional changes
  - Update JVM options to build
  - Update SBT to 1.1.2
- Add experimental `ObjectTree` pretty printer
  - As documented, this is not suitable for use beyond debugging.

## 4.0

- Drop support for Java 6
- Add `skipNones` support in `ObjectTree`. It remains experimental
- Dependency updates (binary compatible)
  - Update Akka to 2.5.12
  - Update Twitter Util to 18.4.0
  - Update Liquibase to 3.6.1

#### 4.0.1

- Improvements to `ObjectTree`
  - Fix an inefficiency in sequence traversal
  - Fix a bug when `skipNones` was `false`
  - Add a maximum recursion depth
  - Add a length limit for single-line 1-products

### 4.1

- Update Scala to 2.12.6
- Update SBT to 1.1.6
- Tweak `ObjectTree` placeholder string for too-deep recursion
- Dependency updates (binary compatible)
  - Updated Akka to 2.5.13
  - Updated Akka HTTP to 10.1.3
  - Updated Json4s to 3.5.4
  - Updated Twitter Util to 18.6.0
  - Updated Scala Parser Combinators to 1.1.1
  - Updated Joda Time to 2.10
  - Updated Liquibase to 3.6.2
- Add `WrappedPreparedStatement` and `WrappedResultSetMetadata`

### 4.2

- Update Scala to 2.12.8
- Update SBT to 1.2.8
- Add Metals & Bloop to build
- Dependency updates
  - Updated Akka to 2.5.23
  - Updated SLF4J to 1.7.26
  - Updated Log4s to 1.8.2
  - Updated Joda Time to 2.10.2
  - Updated Joda Convert to 2.2.1
  - Updated ThreeTenBP to 1.4.0
  - Updated Commons VFS to 2.3
  - Updated Groovy to 2.5.7
  - Updated Json4s to 3.6.7
  - Updated Scala Parser Combinators to 1.1.2
  - Updated Scala XML to 1.2.0
  - Updated BouncyCastle to 1.62
  - Updated Liquibase to 3.6.3
  - Updated Slick to 3.3.2
  - Updated Scalatest to 3.0.8

## 5.0

- Remove `gerweck-utils-twitter`
  - This is supported in Twitter's bijection library
- Switch from `org.spire-math:spire:0.13.0` to `org.typelevel:spire:0.16.2`
- Add new `BuildingFlow` in Akka utils
- Add Scala 2.13.0

project/Build.scala
import sbt._
import Keys._

import sbtrelease.ReleasePlugin.autoImport._
import com.typesafe.sbteclipse.plugin.EclipsePlugin._

import Helpers._

object UtilsBuild extends Build {
  build =>

  import BuildSettings._
  import Resolvers._
  import Dependencies._
  import PublishSettings._
  import Helpers._

  lazy val basicLogDeps = Seq(
    slf4j,
    log4s,
    logback % "test",
    groovy % "test"
  )

  lazy val utilsDeps = basicLogDeps ++ Seq (
    jclBridge,
    scalaCheck % "test",
    commonsIo,
    jodaTime % "optional",
    jodaConvert % "optional",
    /* ThreeTen is optional in some versions and not others, so see below */
    spire % "provided,optional",
    commonsVfs
  )

  lazy val macros = (project in file ("macro"))
    .commonSettings()
    .settings(Eclipse.settings: _*)
    .settings(falsePublishSettings: _*)
    .settings (
      name := "Gerweck Utils Macros",
      libraryDependencies += log4s,
      libraryDependencies <+= scalaVersion("org.scala-lang" % "scala-reflect" % _),
      resolvers += Resolver.sonatypeRepo("releases"),

      scalaSource in Compile := {
        scalaBinaryVersion.value match {
          case "2.10" => baseDirectory.value / "src" / "main" / "scala-2.10"
          case _      => baseDirectory.value / "src" / "main" / "scala-2.11"
        }
      },

      publish := {},
      publishLocal := {},
      publishArtifact := false,

      exportJars := false,
      releaseProcess := Seq.empty
    )

  lazy val root: Project = (project in file ("."))
    .aggregate(macros, core, java6, twitter, akka, dbutil)
    .moduleSettings()
    .settings (
      name := "Gerweck Utils Root",

      publish := {},
      publishLocal := {},
      publishArtifact := false,

      exportJars := false,

      skip in Compile := true,
      skip in Test := true
    )

  lazy val core: Project = (project in file ("core"))
    .dependsOn(macros % "optional")
    .moduleSettings()
    .settings(
      name := "Gerweck Utils",

      scalacOptions ++= scalacOpts(SVer(scalaBinaryVersion.value), true),
      javacOptions ++= javacOpts(true),

      libraryDependencies ++= utilsDeps,
      libraryDependencies += threeTen % "optional",
      libraryDependencies += scalaTest(scalaBinaryVersion.value) % "test",
      libraryDependencies <+= scalaVersion("org.scala-lang" % "scala-reflect" % _),

      libraryDependencies <++= (scalaBinaryVersion) (scalaParser),
      libraryDependencies <++= (scalaBinaryVersion) (scalaXml),

      resolvers += Resolver.sonatypeRepo("releases"),

      unmanagedSourceDirectories in Compile <+= (scalaBinaryVersion, baseDirectory) { (ver, dir) =>
        ver match {
          case "2.10" => dir / "src" / "main" / "scala-2.10"
          case _      => dir / "src" / "main" / "scala-2.11"
        }
      },

      unmanagedSourceDirectories in Compile += baseDirectory.value / "src" / "main" / "scala-java8",

      // include the macro classes and resources in the main jar
      mappings in (Compile, packageBin) ++= mappings.in(macros, Compile, packageBin).value,

      // include the macro sources in the main source jar
      mappings in (Compile, packageSrc) ++= mappings.in(macros, Compile, packageSrc).value,

      // Do not include macros as a dependency.
      pomPostProcess := excludePomDeps { (group, artifact) => (group == "org.gerweck.scala") && (artifact startsWith "gerweck-utils-macro") }
    )

  lazy val java6 = (project in file ("java6"))
    .dependsOn(macros % "optional")
    .moduleSettings()
    .settings(
      name := "Gerweck Utils (Java 6)",
      normalizedName := "gerweck-utils-java6",

      scalacOptions ++= scalacOpts(SVer(scalaBinaryVersion.value), false),
      javacOptions ++= javacOpts(false),

      libraryDependencies ++= utilsDeps,
      libraryDependencies += threeTen,
      libraryDependencies += scalaTest(scalaBinaryVersion.value) % "test",
      libraryDependencies <+= scalaVersion("org.scala-lang" % "scala-reflect" % _),

      libraryDependencies <++= (scalaBinaryVersion) (scalaParser),
      libraryDependencies <++= (scalaBinaryVersion) (scalaXml),

      resolvers += Resolver.sonatypeRepo("releases"),

      unmanagedSourceDirectories in Compile <+= (scalaBinaryVersion, baseDirectory) { (ver, dir) =>
        ver match {
          case "2.10" => dir / ".." / "core" / "src" / "main" / "scala-2.10"
          case _      => dir / ".." / "core" / "src" / "main" / "scala-2.11"
        }
      },
      scalaSource in Compile := baseDirectory.value / ".." / "core" / "src" / "main" / "scala",
      scalaSource in Test := baseDirectory.value / ".." / "core" / "src" / "test" / "scala",
      unmanagedSourceDirectories in Compile += baseDirectory.value / ".." / "core" / "src" / "main" / "scala-java6",

      // include the macro classes and resources in the main jar
      mappings in (Compile, packageBin) ++= mappings.in(macros, Compile, packageBin).value,

      // include the macro sources in the main source jar
      mappings in (Compile, packageSrc) ++= mappings.in(macros, Compile, packageSrc).value,

      // Do not include macros as a dependency.
      pomPostProcess := excludePomDeps { (group, artifact) => (group == "org.gerweck.scala") && (artifact startsWith "gerweck-utils-macro") }
    )

  lazy val twitter = (project in file ("twitter"))
    .moduleSettings()
    .settings(
      name := "Gerweck Utils Twitter",
      libraryDependencies ++= basicLogDeps,
      libraryDependencies += twitterUtil % "optional"
    )

  lazy val akka: Project = (project in file ("akka"))
    .dependsOn(core)
    .moduleSettings()
    .settings(
      name := "Gerweck Utils Akka",

      scalacOptions ++= scalacOpts(SVer(scalaBinaryVersion.value), true, false),
      javacOptions ++= javacOpts(true),

      /* Logging */
      libraryDependencies ++= basicLogDeps,
      /* Akka */
      libraryDependencies ++= Seq (
        akkaActor,
        akkaStream
      )
    )

  lazy val dbutil: Project = (project in file ("dbutil"))
    .dependsOn(core)
    .moduleSettings()
    .settings(
      name := "Gerweck Utils DB",

      scalacOptions ++= scalacOpts(SVer(scalaBinaryVersion.value), true),
      javacOptions ++= javacOpts(true),

      /* Slick dependencies */
      libraryDependencies ++= Seq (
        slick
      ),
      /* Optional mappings */
      libraryDependencies ++= Seq (
        json4sNative % "optional",
        akkaHttpCore % "optional"
      )
    )

  private implicit class ProjectHelper2(p: Project) {
    def moduleSettings() = {
      p .commonSettings
        .settings(Eclipse.settings: _*)
        .settings(EclipseKeys.skipParents in ThisBuild := false)
        .settings(publishSettings: _*)
        .settings(Release.settings: _*)
    }
  }
}
